File: src/Medium/Js/TopVotedCandidate.js
/**
 * @param {number[]} persons
 * @param {number[]} times
 */
var TopVotedCandidate = function (persons, times) {
    let max = 0;
    let v = {}, winner = null;
    this.times = times;
    this.winMap = times.reduce((pre, cur, index) => {
        let p = persons[index];
        v[p] =  (v[p] || 0) + 1;
        let n = v[p];
        // console.log(p,n,max,winner);
        if (n >= max || winner===null) {
            max = n;
            winner = p;
        }
        pre[cur] = winner;
        return pre;
    }, {})
};

/** 
 * @param {number} t
 * @return {number}
 */
TopVotedCandidate.prototype.q = function (t) {
    let l = 0, r = this.times.length - 1;
    while (l < r) {
        let mid = ~~((r + l) / 2);
        if (t < this.times[mid]) {
            r = mid - 1;
        } else if (t > this.times[mid]) {
            l = mid +1;
        } else {
            l = mid;
            break;
        }
    }
    if(this.times[l]>t&&l>0) l--;
    let index = this.times[l];
    // console.log(l,index);
    return this.winMap[index];
};


 // Your TopVotedCandidate object will be instantiated and called as such:
  var obj = new TopVotedCandidate([0,1,0,1,1],[24,29,31,76,81])
  console.log(obj.winMap);
//   var param_1 = obj.q(t)
 console.log(obj.q(25));
//  [[],[45],[49],[59],[68],[42],[37],[99],[26],[78],[43]]

File: src/Medium/DesignTest/TopVotedCandidate.py
from typing import List, Tuple, Union
from heapq import heappop, heappush, heapify, heappushpop, heapreplace
from collections import defaultdict, deque, Counter
from itertools import accumulate, permutations, combinations, product, compress
from math import perm, comb, gcd, lcm
from functools import cache, lru_cache, reduce
from sortedcontainers import SortedList, SortedSet, SortedDict
from bisect import bisect_left, bisect_right


class TopVotedCandidate:

    def __init__(self, persons: List[int], times: List[int]):
        self.persons = persons
        self.times = times
        buc = defaultdict(int)
        max_vote, p = 0, 0
        self.tp = defaultdict(int)
        for person, time in zip(persons, times):
            buc[person] += 1
            if buc[person] >= max_vote:
                max_vote = buc[person]
                p = person
            self.tp[time] = p

    def q(self, t: int) -> int:
        return self.tp[self.times[bisect_right(self.times, t) - 1]]


# Your TopVotedCandidate object will be instantiated and called as such:
# obj = TopVotedCandidate(persons, times)
# param_1 = obj.q(t)


if __name__ == '__main__':
    s = TopVotedCandidate([0, 1, 1, 0, 0, 1, 0], [0, 5, 10, 15, 20, 25, 30])
    print(s.q(3))
